### ISSUE TITLE
Poor download speed with cache_size = 0

### ISSUE THREAD

**Issue by ssiloti** [arvidn/libtorrent#527]
When running RC_1_1 with the disk cache disabled I've noticed libtorrent sometimes gets into a state where downloading slows to a crawl for up to several minutes. I think this is what's happening: After peers allocate their first disk buffer to receive piece data they all get their `download_channel` flagged with `bw_disk` due to exceeding the max cache size of zero. As the writes are retired and the buffers freed `check_buffer_level()` is called, but it always bails out because there are at least one-or-two buffers allocated, so `m_in_use > m_low_watermark`. Eventually a huge number of connections are blocked waiting for `peer_connection::on_disk()` to clear `bw_disk` while a trickle of new peers download a single piece block before joining the queue. At some point `check_buffer_level()` gets lucky and is called while there are no buffers allocated which triggers a huge burst of download activity.

I'm still trying to think of a good solution to this. The obvious fix would be to enforce a minimum cache size, but it seems like there should be a better way.

**Comment by arvidn** [comment]
I've noticed the slowdown myself while debugging issues around the cache. your analysis sounds right.
I imagine guaranteeing that each connection can have at least one block allocated at a time (or maybe even 2) would solve it. I'm not sure what the best way of implementing such guarantee though

**Comment by ssiloti** [comment]
It seems to me that the problem we're dealing with here is very similar to scheduling packets in a router. We're looking for the smallest buffer size which will maintain maximum disk utilization for any given network ingress pattern. Perhaps we can borrow some ideas from that realm, like using the amount of time write jobs sit in the queue to decide when to set `exceeded` to true and to signal `peer_connection::on_disk()`. As long as write jobs are being retired quickly we can freely allow new receive buffers to be allocated, but if they start sitting around for a while that probably means the disk IO threads are blocking waiting for the OS to flush pages from its cache. When that happens we want to propagate back-pressure to the network via `exceeded` until the delay drops below some nominal value.

**Comment by ssiloti** [comment]
I made a quick and dirty [proof-of-concept](https://github.com/ssiloti/libtorrent/commit/c296a5e5cfd95338bdff26e30a8036efbc0dcd92) to experiment with this. I've put it on hold for the moment while I look at #530.

**Comment by arvidn** [comment]
if I understand your patch correctly, as long as any single write operation takes less than a second, no peer will stay blocked by the disk. Maybe I'm missing something, but that would not prevent the disk write queue to grow indefinitely, right? unless the timing includes the queue time of the block (but I don't think it does, does it?).

more fundamentally though, I would be a bit concerned about heuristics based on timing, because I would expect completion times of writes to vary a lot, and come and go in bursts.

I believe there is a clear trade-off between the amount of memory that will be used and the download performance. If you aren't willing to use enough, your download rate will suffer. I'm not entirely convinced that's such a bad thing. Obviously the way it works now is really bad and need fixing, but I'm not sure the solution needs to prioritize performance (and bend the cache size configuration to accommodate).

I think it would be simpler, and perhaps even more expected behavior, to violate the cache setting as little as possible to avoid the degenerate state you discovered, but no more. With that in mind, I'm inclined to add a simple rule saying:

"every peer may have 2 allocated disk blocks at any given time, regardless of whether the cache says it's full or not"

I think that would avoid the problem, probably not provide the best download performance per memory usage (because the memory usage won't be balanced by peer capacity for instance), but I think it would be simple robust and "good enough".

what do you think? I'll write a patch to test.

**Comment by arvidn** [comment]
btw, the same problem presumably exist with cache_size = 1 and = 2 as well.. and any other low cache_size setting (meaning, it's not safe to enable a different heuristic when cache_size == 0 to solve this).

**Comment by arvidn** [comment]
as far as I can tell, this solves the problem (running with cache_size = 1, as you noticed, cache_size = 0 causes incorrect hash failures).

https://github.com/arvidn/libtorrent/pull/534

**Comment by ssiloti** [comment]
The threshold for throttling is 1 millisecond, the timestamps are in microseconds. The timing does include the queue time. Basically this is half-assed CoDel.

Backing up a bit, let me explain where I'm coming from. I don't want libtorrent's cache enabled because it mostly just wastes memory duplicating data which is already in the OS's page cache. This would be especially true if libtorrent used mmap for IO, which I know it doesn't, but it should. To me dirty buffers which are queued to be written aren't really cached, they're part of a rate matching buffer between the network and the disk.

While it does look like #534 fixes the immediate problem, it'd be nice to have a mode which neither wastes memory by double caching nor sacrifices performance when there are still plenty of resources available.

**Comment by arvidn** [comment]
I agree that in the long term libtorrent should use (or at least support) mmapped disk I/O.

The main purpose (as far as I understand) the application-level dirty buffers serve is to essentially pin them in RAM until we've run the sha-1 hash passed them. If they get flushed to the page cache, even if they're not flushed and evicted by the kernel, it will be somewhat expensive to retrieve them back into user space to hash them (mostly because of the system call and memory copy).

In my mind, the next obvious step is to significantly simplify the disk_buffer_pool to allocate its block out of anonymously mmapped memory. an extension of that would be to allow allocating blocks that are directly backed by their target file on disk. There are a few real challenges with that though:

1. when a disk is full, surprisingly recent versions of linux allocate physical memory pages despite not being able to allocate the corresponding page on disk, and the data will silently be discarded once the kernel realize the disk is full.

2. On recent versions of linux, page faulting a page that won't fit on disk will cause a segfault. This can be quite complicated to handle and do anything sensible with. Basically, each 4 kiB page of a 16 kiB allocation would have to be touched and page faulted in the disk thread, to make sure they can be allocated on disk, before returning the allocation to the network thread.

3. It's not obvious that memory mapped I/O would perform very well with torrents that have a lot of small files.

4. Not all systems support memory mapped I/O, so libtorrent would probably still need to retain a fallback mechanism to regular disk I/O, probably to mitigate (3) as well.

If you have any thoughts of ways of solving or simplifying any of those, I would be excited to hear about it!

**Comment by ssiloti** [comment]
Regarding out-of-space handling. On linux you can use fallocate(2) to efficiently pre-allocate space for a file. In fact I see that libtorrent is already using it via the standard function posix_fallocate(). As long as you're using posix_fallocate before you try to write you shouldn't be seeing ENOSPC on a page fault.

Why would mmapped IO be particularly slow for small files? At worst it's a page fault for every file which I'm pretty sure is not significantly worse than a syscall for every file as is required with write().

Are there really enough systems out there without mmap to justify carrying around such complex code? I can't think of any such platforms which would account for more than a fraction of a percent of potential libtorrent users.

**Comment by arvidn** [comment]
(3) is pure speculation. I primarily get this impression from people benchmarking libtorrent against rtorrent and finding that it performed worse, even for torrents that would completely fit in RAM. Now, it could be the case that rtorrent does not recognize that mmapped file I/O is still *blocking* I/O, and that it should be done in other threads than the main network loop.

But basically the thought would be that mmapping() and munmapping() and basically managing a fragmented memory space would be costly. It's not just the system calls that cost, but also the TLB miss/page fault and allocating a fresh page. Now, that may happen in the kernel for a read() call too though. I did recently find that when checking the piece hashes of a large torrent, reusing disk buffers to read the file into was noticeably quicker than to read into new disk buffers (allocated but presumably not committed/assigned physical memory yet by the kernel).

perhaps preallocating space would be the way to go. there's a risk that may cause usability challenges too, with users that expect to be able to use sparse files. (it's certainly a pain on OSX sometimes that HFS+ does not support sparse files). It would also mean having to be a lot more careful with deselected files, and probably to support transitioning bytes back and forth between the partfile and the actual files.

In order to make sure the actual disk I/O doesn't block the network thread though, I would imagine a disk thread would have to make sure it causes all necessary page faults first, blocking that thread. Both when reading and writing (I imagine that's a matter of msync() and MADV_WILLNEED + reading the first byte of each page).

**Comment by arvidn** [comment]
addressed in: https://github.com/arvidn/libtorrent/pull/545

**Comment by ssiloti** [comment]
I started a test torrent this morning using 5d499e83b698b3fbf782e97e023fb8ab00ec2aa7 (latest RC_1_1) with cache_size set to zero and saw the same download speed collapse again. After taking a second look at 08bac479 I realized m_outstanding_bytes isn't quite suitable for what that commit is trying to use it for. m_outstanding_bytes isn't the number of bytes outstanding to the disk like I assumed, it's the number of bytes requested from the remote peer. This means that while it is less likely a peer will assert bw_disk, once it does it is stuck in the same hang as before because writes completing at the disk do not decrement m_outstanding_bytes, receiving piece data does, and that is precisely what is being blocked by bw_disk being asserted.
